src/clj/rems/api/services/dependencies.clj

```diff
@@ -0,0 +1,131 @@
+;; The :archived and :enabled bits are not stored in the graph because
+;; that would mean invalidating the whole graph on every status bit
+;; change, _or_ implementing partial rebuilding.
+;;
+;; TODO: is this caching necessary? Computing the dependencies takes
+;; 500ms with performance test data.
+(def ^:private dependencies-cache (atom nil))
+
+(defn dependencies []
+  (when-not @dependencies-cache
+    (reset! dependencies-cache (compute-dependencies)))
+  @dependencies-cache)
+
+(defn reset-cache! []
+  (reset! dependencies-cache nil))
+
+;; TODO change format of errors so we can get rid of this conversion
+(defn- format-deps [deps]
+  (apply merge-with concat
+         (for [dep deps]
+           (cond (:license/id dep)
+                 {:licenses [(select-keys (enrich-dependency dep) [:id :localizations])]}
+
+                 (:resource/id dep)
+                 {:resources [(select-keys (enrich-dependency dep) [:id :resid])]}
+
+                 (:workflow/id dep)
+                 {:workflows [(select-keys (enrich-dependency dep) [:id :title])]}
+
+                 (:catalogue-item/id dep)
+                 {:catalogue-items [(select-keys (enrich-dependency dep) [:id :localizations])]}
+
+                 (:form/id dep)
+                 {:forms [(select-keys (enrich-dependency dep) [:form/id :form/title])]}))))
+
+(defn archive-errors
+  "Return errors if given item is depended on by non-archived items"
+  [error-key item]
+  (when-let [users (->> (get-in (dependencies) [:reverse-dependencies item])
+                        (mapv add-status-bits)
+                        (remove :archived)
+                        seq)]
+    [(merge {:type error-key}
+            (format-deps users))]))
+
+(defn in-use-errors
+  "Returns errors if given item is depended on at all"
+  [error-key item]
+  (when-let [users (seq (get-in (dependencies) [:reverse-dependencies item]))]
+    [(merge {:type error-key}
+            (format-deps users))]))
+
+(defn unarchive-errors
+  "Return errors if given item depends on archived items"
+  [item]
+  (when-let [used (->> (get-in (dependencies) [:dependencies item])
+                       (mapv add-status-bits)
+                       (filter :archived)
+                       seq)]
+    (let [{:keys [licenses resources workflows catalogue-items forms]} (format-deps used)]
+      (concat
+       (when licenses
+         [{:type :t.administration.errors/license-archived
+           :licenses licenses}])
+       (when resources
+         [{:type :t.administration.errors/resource-archived
+           :resources resources}])
+       (when workflows
+         [{:type :t.administration.errors/workflow-archived
+           :workflows workflows}])
+       (when forms
+         [{:type :t.administration.errors/form-archived
+           :forms forms}])
+       ;; case not possible:
+       (when catalogue-items
+         [{:type :catalogue-items-archived
+           :catalogue-items catalogue-items}])))))
```

Comment:
This feels like code that can perhaps be removed? At least having separate cases for all the different types could perhaps be removed by restructuring the errors. How about we had a common `:type` for all archived errors? The items themselves could tell which kinds were wrong. I think the UI will just display a list of all items and does not care about grouping by types.

Reply:
I had a TODO about that but ended up not doing it for some reason. It's a good idea.

Reply:
as discussed, I'll do this in a follow-up PR